SYSDB: perf improvements in `sysdb_add_group_member_overrides()`

alexey-tikhonov opened the pull request:
Resolves: https://github.com/SSSD/sssd/issues/7793

alexey-tikhonov commented:
Ok, so I've conducted some testing.

Setup:
 - 50k users with own primary groups + `final-g5050002` group that contains all those users.
 - ldap-provider, rfc2307bis; debug_level = 1; debug_backtrace_enabled = false
 - everything is cached: `getent group final-g5050002` + `id` every member.

Testing on a laptop with F42 in a container with:
 - **stock**: sssd-2.10.2-2
 - **reduced**: this PR minus last patch (`sysdb_asq_search()`) instead of search-by-memberof)
 - **full**: this PR

`time SSS_NSS_USE_MEMCACHE=NO getent -s sss group final-g5050002@ldap.test > /dev/null`

Results:
|                 | mode: power saver             | mode: performance            |
| ------------ | -------------------------------------- | ------------------------------------- |
| stock      | 2.203 .. 2.399 sec                  | 0.930 .. 0.955 sec                 |
| reduced | 1.806 .. 1.881 sec (~ -20%)  | 0.761 .. 0.782 sec (~ -18%)  |
| full          | 1.268 .. 1.384 sec (~ -42%)  | 0.541 .. 0.545 sec (~ -42%)  |

Take a note that in a more realistic setup - a bunch of groups containing modest subset of users (instead of all users) - an effect of "reduced" patch set will be much less pronounced, but effect of "`sysdb_asq_search` vs search-by-memberof" depends more on entire db size and less on the group size.

alexey-tikhonov commented:
Additional feedback from @marco-kusa , reporter of #7793 :
"id user" time in their environment, with cache fully populated:
 - with the stock package with additional debug (with `ignore_group_members` both true and false): `real  7m16.568s`
 - with all those patches applied and `ignore_group_members = false` is significantly improved: `real  0m29.152s`
 - with all those patches applied and `ignore_group_members = true` is almost instant: `real  0m0.221s`

alexey-tikhonov commented:
I've removed 'sssd-2-9' label because currently there is a conflict in tests.
But in general my proposal is:
 - backport all functional (product) patches to 'sssd-2-9'
 - backport only ` SYSDB: don't add group members if 'ignore_group_members == true' ` to 'sssd-2-9-4'

joakim-tjernlund commented:
Added these patches at work and yes, lookups are MUCH faster!
Tested enumerate=true just for kicks and that worked too.

pbrezina commented:
Hi, others will review the patches, but I'm curious. Can you write down how did you improve the performance so dramatically?

alexey-tikhonov commented:
Hi @pbrezina,

> Can you write down how did you improve the performance so dramatically?

First off all to stress: this is only about hot, fully populated and huge cache. I.e. about cache read operations, specifically `getgrnam()`/`getgrgid()`.
Inspection of logs with microseconds enabled shows that **vast** majority of time in those cases is currently spent in `sysdb_add_group_member_overrides()`.

Then it's two-fold:

(1) 'ignore_group_members = true' case.
That's quite trivial - b842b952a63c39958f7c5e2a18c7fa9827c81bbd - just don't bother with group members, and hence member overrides, at all (and I generalized this a bit, but important part is really a couple of lines in `sysdb_add_group_member_overrides()`). And lookup gets instant (take a note that all the work done was wasted because members aren't returned in this case anyway).

(2) 'ignore_group_members = false' case.
1) Main thing here is a change to use `sysdb_asq_search()` (dereference of readily available 'members' list - thanks Sumit for the function name, I couldn't find it in the docs/headers) instead of blunt "(&("SYSDB_UC")("SYSDB_MEMBEROF"=%s))" search - 9ea35c2d955d7522e611e2feec65972afbe3ca56.
Effect of this change is very pronounced when the cache contains a lot of users (think of 10..100k) and group contains a small subset (less than 1k): in this case dereference is **much** faster.
2) Less important things are tiny optimizations in a loop that iterates over resulting list: just reducing the number of strcpy(), memcpy(), etc. This is what I called "reduced" in https://github.com/SSSD/sssd/pull/7841#issuecomment-2675248906 While being tiny they still bring around 20% gain in my testing, so not negligible.

aplopez commented on src/db/sysdb_views.c:
```diff
@@ -1688,7 +1687,7 @@ errno_t sysdb_add_group_member_overrides(struct sss_domain_info *domain,
         }
 
-        val = talloc_strdup(obj, memberuid);
+        val = talloc_steal(obj, memberuid);
```
Is it safe to steal `memberuid`?
In the case where `memberuid` was initially `NULL`, it's OK because it will be allocated by `add_domain_name()` attached to `tmp_ctx`. But when `memberuid` is not initially `NULL`, `memberuid` is assigned `orig_name` which was retrieved by `ldb_msg_find_attr_as_string()` and it returns an internal value (not a copy) which, I guess, is attached to the `msg`.

alexey-tikhonov replied:
> But when `memberuid` is not initially `NULL`, `memberuid` is assigned `orig_name` 

It's the other way around: "memberuid = orig_name" when override not found, i.e. "memberuid == NULL".

> which was retrieved by `ldb_msg_find_attr_as_string()` and it returns an internal value (not a copy) which, I guess, is attached to the `msg`.

Note that in all code paths `memberuid` is set to a pointer returned by `ldb_msg_find_attr_as_string()`, just different obj/attr is searched.

`talloc_steal()` will re-attach to a new context ('obj'). So when `talloc_free(tmp_ctx)` frees 'res_members'/'override_obj', this value should be left intact.

Do I miss something?

aplopez replied:
OK. I had missed that `override_obj` was attached to `res_members`, which was in turn attached to `tmp_ctx`, which is freed at the end of the function.

alexey-tikhonov commented:
@aplopez,
@sumit-bose has a valid point that straightforward replacement of "search-by-memberof" with "deref(members)" won't work in case of nested groups because "members" contain only direct members.
For this reason I has withdrawn this patch and will try to amend it in a separate PR.

I also dropped tests related patches for reasons described above and rebased.

sumit-bose reviewed (approved):
Hi,

thank you for the update, I didn't find any other issues, ACK.

bye,
Sumit

alexey-tikhonov commented:
Pushed PR: https://github.com/SSSD/sssd/pull/7841

* `master`
    * ed6956e92d9b034ad774f3c2c090d86b7d18769f - SYSDB: update in get_user_members_recursively()
    * f61b9bbb2d918f948f7bd60c8c541a9ead98bf37 - SYSDB: make `sysdb_get_user_members_recursively()` static
    * 0a9ae2c2b23c003d3900048b20a7bc767784bf24 - SYSDB: update in sysdb_add_group_member_overrides()
    * 6aae3572af88e1ade4965d02387a50747030a7a7 - SYSDB: update in sysdb_add_group_member_overrides()
    * 108800dc9a42355c32fa6031d1b2c0bd4b7dc8ec - SYSDB: update in sysdb_add_group_member_overrides()
    * a58aa915f7354e1b006978e8ed829b8cf2c860d2 - SYSDB: update in sysdb_add_group_member_overrides()
    * 6c50506c4361d3f8672d1481c77a221212f23e39 - SYSDB: fix sysdb_add_group_member_overrides()
    * 6e01e4127eb762fcfb06bc5cf9aefd53359e3825 - SYSDB: update in sysdb_add_group_member_overrides()
    * 501663f2a54750d7a63fc0b48ac4f878060459fb - SYSDB: update in sysdb_add_group_member_overrides()
    * 5e882b3660b733aca698ebea6999273b5a59e49e - SYSDB: update in sysdb_add_group_member_overrides()
    * 281d9c3ed66ee28a9572433a629eb0d72525ca46 - SYSDB: don't add group members if 'ignore_group_members == true'
* `sssd-2-9`
    * b64d205c0ceff381d87a2d4d310d9a9ca77c3051 - SYSDB: update in get_user_members_recursively()
    * 66853342491d51889294219cbe3f6002bf699d88 - SYSDB: make `sysdb_get_user_members_recursively()` static
    * 02a5995303ccd514f64603ac8901ed24438064e9 - SYSDB: update in sysdb_add_group_member_overrides()
    * 3477221b9b91807dd67ae1351959c30ec8fca899 - SYSDB: update in sysdb_add_group_member_overrides()
    * c4100177b9490b3896e12c3f439a1b148b109dfb - SYSDB: update in sysdb_add_group_member_overrides()
    * 638b904c1ad58806d4b3a185c0bd63ccac50d52c - SYSDB: update in sysdb_add_group_member_overrides()
    * 6484049649f5ddeec4f778b4e18962f42374ea00 - SYSDB: fix sysdb_add_group_member_overrides()
    * f7964c9929d899f105e233190700efa1b893b515 - SYSDB: update in sysdb_add_group_member_overrides()
    * ccc90e2fbda51bfcd0356b0825fbacb3e2057851 - SYSDB: update in sysdb_add_group_member_overrides()
    * 73f6c4a70eed5fa84d20a62cab7a406a0abb42c2 - SYSDB: update in sysdb_add_group_member_overrides()
    * addb1a78106cab8a85f8f6c56d79e84b5abd0d5e - SYSDB: don't add group members if 'ignore_group_members == true'

alexey-tikhonov commented:
> For this reason I has withdrawn this patch and will try to amend it in a separate PR.

#7866
